File: src/cloudproxy/file_client.cc
//  File: file_client.cc
//  Author: Tom Roeder <tmroeder@google.com>
//
// Description: Implementation of the FileClient class that interacts
// with FileServer
//
//
//  Copyright (c) 2013, Google Inc.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
#include <string>

#include <glog/logging.h>
#include <keyczar/keyczar.h>

#include "cloudproxy/cloud_auth.h"
#include "cloudproxy/cloud_user_manager.h"
#include "cloudproxy/file_client.h"
#include "tao/tao_auth.h"

using std::string;

namespace cloudproxy {

FileClient::FileClient(const string &file_path,
                       const string &client_config_path,
                       tao::TaoChildChannel *channel, tao::TaoDomain *admin)
    : CloudClient(client_config_path, channel, admin), file_path_(file_path) {}

bool FileClient::Init() {
  if (!CloudClient::Init()) {
    LOG(ERROR) << "Could not initialize file cloud client";
    return false;
  }
  struct stat st;
  if (stat(file_path_.c_str(), &st) != 0) {
    LOG(ERROR) << file_path_ << " does not exist";
    return false;
  }
  if (!S_ISDIR(st.st_mode)) {
    LOG(ERROR) << file_path_ << " is not a directory";
    return false;
  }
  return true;
}

bool FileClient::Create(SSL *ssl, const string &owner,
                        const string &object_name) {
  // defer to the CloudClient implementation to get this created, since there's
  // nothing else to do with the file
  return CloudClient::Create(ssl, owner, object_name);
}

bool FileClient::Destroy(SSL *ssl, const string &owner,
                         const string &object_name) {
  // defer to the CloudClient implementation to get this destroyed, since
  // there's nothing else to do with the file
  return CloudClient::Destroy(ssl, owner, object_name);
}

bool FileClient::Read(SSL *ssl, const string &requestor,
                      const string &object_name, const string &output_name) {
  // make the call to get permission for the operation, and it that succeeds,
  // start to receive the bits
  CHECK(CloudClient::Read(ssl, requestor, object_name, output_name))
      << "Could not get permission to READ " << object_name;

  string path = file_path_ + string("/") + output_name;
  CHECK(ReceiveStreamData(ssl, path)) << "Error while reading the"
                                      << " file and writing it to disk";
  return HandleReply(ssl);
}

bool FileClient::Write(SSL *ssl, const string &requestor,
                       const string &input_name, const string &object_name) {
  // look up the file to get its length and make sure there is such a file
  string path = file_path_ + string("/") + input_name;
  struct stat st;
  CHECK_EQ(stat(path.c_str(), &st), 0) << "Could not stat the file " << path;

  // make the call to get permission for the operation, and if that succeeds,
  // then start to write the bits to the network
  CHECK(CloudClient::Write(ssl, requestor, input_name, object_name))
      << "Could not get permission to write to the file";

  VLOG(2) << "Got permission to write the file " << path;

  CHECK(SendStreamData(path, st.st_size, ssl))
      << "Could not send the file data to the server";
  return HandleReply(ssl);
}

}  // namespace cloudproxy

File: go/apps/newfileproxy/fileclient/fileclient.go
// Copyright (c) 2016, Google, Inc.,  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//     http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// File: fileclient.go

package main

import (
	"bytes"
	"crypto/rand"
	"crypto/x509"
	"flag"
	"fmt"
	"io/ioutil"
	"path"

	"github.com/golang/protobuf/proto"
	"github.com/jlmucb/cloudproxy/go/taotataoo"
	taosupport "github.com/jlmucb/cloudproxy/go/support_libraries/tao_support"

	"github.com/jlmucb/cloudproxy/go/apps/newfileproxy/common"
	"github.com/jlmucb/cloudproxy/go/apps/newfileproxy/resourcemanager"
)

var caAddr = flag.String("caAddr", "localhost:8124", "The address to listen on")
var simpleCfg = flag.String("domain_config",
	"./tao.config",
	"path to tao configuration")
var fileClientPath = flag.String("path",
	"./FileClient",
	"path to FileClient files")
var serverHost = flag.String("host", "localhost", "address for client/server")
var serverPort = flag.String("port", "8123", "port for client/server")
var useSimpleDomainService = flag.Bool("use_simpledomainservice", true,
	"whether to use simple domain service")
var serverAddr string

func main() {

	// This holds the cloudproxy specific data for fileclient
	// including the Program Cert and Program Private key.
	var clientProgramData taosupport.TaoProgramData

	// Make sure we zero keys when we're done.
	defer clientProgramData.ClearTaoProgramData()

	// Parse flags
	flag.Parse()
	serverAddr = *serverHost + ":" + *serverPort

	// If TaoParadigm completes without error, clientProgramData contains all the
	// Cloudproxy information needed throughout fileclient execution.
	err := taosupport.TaoParadigm(simpleCfg, fileClientPath, "ECC-P-256.aes128.hmacaes256",
		*useSimpleDomainService, *caAddr, &clientProgramData)
	if err != nil {
		fmt.Printf("fileclient: Can't establish Tao: ", err)
	}
	fmt.Printf("newfileclient: TaoParadigm complete, name: %s\n",
		clientProgramData.TaoName)

	// Fill Client data
	clientData := new(common.ClientData)
	if clientData == nil {
		fmt.Printf("fileclient: bad clientData init ")
		return
	}
	certificate, err := x509.ParseCertificate(clientProgramData.PolicyCert)
	if err != nil {
		fmt.Printf("fileclient: bad ParseCertificate: ", err)
		return
	}
	clientData.PolicyCert = certificate
	// initialize user keys

	// Get File Secrets
	secretsFileName := path.Join(*fileClientPath, "FileSecrets.bin")

	// fileSecrets is used to encrypt/decrypt client files.
	fileSecrets := make([]byte, 32)

	encryptedFileSecrets, err := ioutil.ReadFile(secretsFileName)
	if err != nil {
		rand.Read(fileSecrets)
	} else {
		fileSecrets, err = tao.Unprotect(clientProgramData.ProgramSymKeys, encryptedFileSecrets)
		if err != nil {
			fmt.Printf("fileclient: Error protecting data\n")
		}
	}

	// Get User Certificates and Private keys
	userKeysFileName := path.Join(*fileClientPath, "serialized_user_keys")
	userKeyFile, err := ioutil.ReadFile(userKeysFileName)
	if err != nil {
		fmt.Printf("fileclient: bad user certs: ", err)
		return
	}
	userKeys := new(common.UserKeysMessage)
	err = proto.Unmarshal(userKeyFile, userKeys)
	if err != nil {
		fmt.Printf("fileclient: bad user certs unmarshal: ", err)
		return
	}

	// Deserialize keys.
	var UserKeyArray []common.KeyData
	for i := 0; i < len(userKeys.SerializedKeys); i++ {
		userKey, err := common.ParseUserKey(userKeys.SerializedKeys[i])
		if err != nil {
		}
		certificate, err := x509.ParseCertificate(userKey.Cert)
		if err != nil {
		}
		userKey.Certificate = certificate
		UserKeyArray = append(UserKeyArray, *userKey)
	}

	// Open the Tao Channel using the Program key. This program does all the
	// standard channel negotiation and presents the secure server name
	// after negotiation is complete.
	ms, serverName, err := taosupport.OpenTaoChannel(&clientProgramData, &serverAddr)
	if err != nil {
		fmt.Printf("fileclient: Can't establish Tao Channel")
		return
	}
	fmt.Printf("fileclient: established Tao Channel with %s, %s\n",
		serverAddr, serverName)

	// Authenticate Principals
	for i := 0; i < len(UserKeyArray); i++ {
		err = common.RequestChallenge(ms, UserKeyArray[i])
		if err != nil {
			fmt.Printf("fileclient: common.RequestChallenge %d fails\n", i)
			return
		}
	}
	fmt.Printf("All common.RequestChallenge's succeeded\n")

	// Create a directory.
	err = common.Create(ms, "directory1", resourcemanager.ResourceType_DIRECTORY, UserKeyArray[0].Cert)
	if err != nil {
		fmt.Printf("fileclient: common.Create 1 fails\n")
		return
	}

	// Create a file.
	err = common.Create(ms, "directory1/file1", resourcemanager.ResourceType_FILE, UserKeyArray[0].Cert)
	if err != nil {
		fmt.Printf("fileclient: common.Create 2 fails\n")
		return
	}
	fmt.Printf("Creates succeeded\n")

	// Add a few owners, readers, writers
	var newcerts [][]byte
	newcerts = append(newcerts, UserKeyArray[1].Cert)
	err = common.AddOwner(ms, "directory1/file1", newcerts)
	if err != nil {
		fmt.Printf("fileclient: common.AddOwner fails\n")
		return
	}
	fmt.Printf("AddOwner succeeded\n")

	newcerts = append(newcerts, UserKeyArray[2].Cert)
	err = common.AddReader(ms, "directory1/file1", newcerts)
	if err != nil {
		fmt.Printf("fileclient: common.AddReader fails\n")
		return
	}
	fmt.Printf("AddReader succeeded\n")

	err = common.AddWriter(ms, "directory1/file1", newcerts)
	if err != nil {
		fmt.Printf("fileclient: common.AddWriter fails\n")
		return
	}
	fmt.Printf("AddWriter succeeded\n")

	// Write a resource.
	file1Contents := []byte{1, 2, 3}
	err = common.WriteResource(ms, "directory1/file1", file1Contents)
	if err != nil {
		fmt.Printf("fileclient: common.WriteResource fails\n")
		return
	}
	fmt.Printf("common.WriteResource succeeded\n")

	// Read a resource.
	recoverdFile1Contents, err := common.ReadResource(ms, "directory1/file1")
	if err != nil {
		fmt.Printf("fileclient: common.ReadResource fails\n")
		return
	}
	if bytes.Compare(file1Contents, recoverdFile1Contents) != 0 {
		fmt.Printf("fileclient: written file differs from read file\n")
	}
	fmt.Printf("common.ReadResource succeeded\n")

	// Tell Server to save state
	err = common.SaveState(ms)
	if err == nil {
		fmt.Printf("common.SaveState succeeded\n")
	} else {
		fmt.Printf("common.SaveState failed\n")
	}

	// Encrypt and store the secret in fileclient's save area.
	encryptedFileSecrets, err = tao.Protect(clientProgramData.ProgramSymKeys, fileSecrets)
	if err != nil {
		fmt.Printf("fileclient: Error protecting data\n")
	}
	err = ioutil.WriteFile(secretsFileName, encryptedFileSecrets, 0666)
	if err != nil {
		fmt.Printf("fileclient: error saving retrieved secret\n")
	}
	fmt.Printf("Secrets and table saved\n")

	// Close down.
	fmt.Printf("fileclient completes with no errors\n")
}
